File: backend/app/services/shipping_id_service.py
import re

from flask import current_app

from app.extensions import db


def generate_shipping_id() -> str:
    from app.models.user import User

    start = current_app.config.get("BOSS_ID_SEQ_START", 90001)
    users = User.query.filter(User.shipping_id.like("BOSS-%")).all()
    max_seq = start - 1
    for user in users:
        match = re.match(r"BOSS-(\d+)$", user.shipping_id)
        if match:
            max_seq = max(max_seq, int(match.group(1)))
    return f"BOSS-{max_seq + 1:05d}"

File: server/src/services/shippingIdService.ts
import { like } from 'drizzle-orm'
import { config } from '../config.js'
import { db } from '../db/index.js'
import { users } from '../db/schema/index.js'

export async function generateShippingId(): Promise<string> {
  const start = config.bossIdSeqStart
  const rows = await db.select({ shippingId: users.shippingId }).from(users).where(like(users.shippingId, 'BOSS-%'))
  let maxSeq = start - 1
  for (const row of rows) {
    const match = /^BOSS-(\d+)$/.exec(row.shippingId)
    if (match) maxSeq = Math.max(maxSeq, parseInt(match[1]!, 10))
  }
  return `BOSS-${String(maxSeq + 1).padStart(5, '0')}`
}
